code/modules/antagonists/changeling/powers/mutations.dm

```diff
@@ -18,28 +18,40 @@
 	var/weapon_name_simple
 
 /datum/action/changeling/weapon/try_to_sting(mob/user, mob/target)
-	if(istype(user.l_hand, weapon_type)) //Not the nicest way to do it, but eh
-		qdel(user.l_hand)
-		if(!silent)
-			user.visible_message("<span class='warning'>With a sickening crunch, [user] reforms [user.p_their()] [weapon_name_simple] into an arm!</span>", "<span class='notice'>We assimilate the [weapon_name_simple] back into our body.</span>", "<span class='warning'>You hear organic matter ripping and tearing!</span>")
-		user.update_inv_l_hand()
-		return
-	if(istype(user.r_hand, weapon_type))
-		qdel(user.r_hand)
-		if(!silent)
-			user.visible_message("<span class='warning'>With a sickening crunch, [user] reforms [user.p_their()] [weapon_name_simple] into an arm!</span>", "<span class='notice'>We assimilate the [weapon_name_simple] back into our body.</span>", "<span class='warning'>You hear organic matter ripping and tearing!</span>")
-		user.update_inv_r_hand()
+	if(istype(user.l_hand, weapon_type) || istype(user.r_hand, weapon_type))
+		retract(user, TRUE)
 		return
 	..(user, target)
 
 /datum/action/changeling/weapon/sting_action(mob/user)
+	SEND_SIGNAL(user, COMSIG_MOB_WEAPON_APPEARS)
```

Comment:
This signal is too specific and is acting more like a proc call. You should just call `retract` in it's place and delete the signal altogether (`try_retract` might be a better name).

Reply:
This signal supposed to support retracting weapons from different datums or even not changeling weapons in future. 
How do i call `retract` or `try_retract` on datum with tentacle while activating armblade?

Reply:
Well your `retract` proc is already generic since it's at the `/datum/action/changeling/weapon` level, and will handle any changeling weapon dynamically since you're using the `weapon_type` var. Simply calling `retract` at the top of the proc should work fine.

Now if you plan to make use of the signal in the future for non-changeling stuff, that's another story. However, I would prefer that it be added in a future when you actually intend to use it.

Reply:
Well uh
Different weapons have different datums, right? Armblade has armblade datum and tentacle has tentacle datum.
If i call `retract` when trying to appear armblade, it will try to retract armblade too, since variable weapon_type on armblade datum is armblade. So it will not work if currently we have tentacle in hand and we're trying to get armblade.

To do this with procs, without signals, we need to somehow access current weapon in hand datum and call its retract function, which is possible but uhh... Why?

Reply:
They are different datums, yes, but they're both children of `/datum/action/changeling/weapon` which means they both run this same `retract` proc. If you have a tentacle out, then click your armblade action, the tentacle would retract, and the armblade would spawn.

Reply:
Actually, no I see what you mean. What I said is wrong.

Reply:
Alright before you wrote that last comment, i tested it on local server and it didnt work confirming that what you said before it is wrong. Marking as resolved